### PR TITLE
[📈⌨Analytics] Build Analytics - Week 3 and 4

### PR DESCRIPTION
**Pull request by qurben**
Week 3 and 4 contributions of the 📈⌨Analytics group.

### CODE REVIEW

File: 03-build-analytics.Rmd [hunk 1]
```diff
@@ -47,12 +62,108 @@ software repositories, open-source software.
 
 ## Answers
 
-<!-- Group per RQ -->
+**RQ1**:
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: Repeat the RQ text in these?

File: 03-build-analytics.Rmd [hunk 2]
```diff
@@ -228,30 +339,41 @@ of around 40% that could be better improved.
...
 
-Version Control Systems (VCS) such as GitHub, and hosted build automation platforms such as Travis, have made Continuous Integration is widely available for projects of every size. This paper suggests that CI is widely used and has improved the quality of processes and developed software itself. However, the article suggests that there is little known about the variety and frequency of errors that cause builds to fail. It suggests that developers should eliminate flaky tests and address common issues regularly such as broken interaction with repositories to keep the build system healthy.
+Version Control Systems (VCS) such as GitHub, and hosted build automation platforms such as Travis,
+have made Continuous Integration is widely available for projects of every size. This paper
+suggests that CI is widely used and has improved the quality of processes and developed software
+itself. However, the article suggests that there is little known about the variety and frequency of
+errors that cause builds to fail. It suggests that developers should eliminate flaky tests and
+address common issues regularly such as broken interaction with repositories to keep the build
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: broken interaction ?
Also, split up the sentence after it.

File: 03-build-analytics.Rmd [hunk 3]
```diff
@@ -228,30 +339,41 @@ of around 40% that could be better improved.
...
+used Travis for their CI services. They also determine that the more popular projects use CI but
+there is no correlation between the popularity of language and usage of CI. It also observes that
+the median project introduces CI a year into development. The paper claims that CI is widely used
+in practise nowadays and CI adoption rates will increase even further in the future. 
 
 ### @rausch2017empirical
 
-Version Control Systems (VCS) such as GitHub, and hosted build automation platforms such as Travis, have made Continuous Integration is widely available for projects of every size. This paper suggests that CI is widely used and has improved the quality of processes and developed software itself. However, the article suggests that there is little known about the variety and frequency of errors that cause builds to fail. It suggests that developers should eliminate flaky tests and address common issues regularly such as broken interaction with repositories to keep the build system healthy.
+Version Control Systems (VCS) such as GitHub, and hosted build automation platforms such as Travis,
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: VCS is used before, but only defined here.

File: 03-build-analytics.Rmd [hunk 4]
```diff
@@ -113,7 +224,7 @@ developers to learn about their experiences in adopting Travis CI.
 
 _Initial Seed_
 
-This paper analyzes what factors have impact on abandonment of Travis. They find that increased
+This paper analyzes what factors have impact on abandonment of Travis\ CI. They find that increased
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: Why Travis\ CI?

Comment (qurben) [reply 03-build-analytics.Rmd]: Not sure where i found `\ ` is nbsp 🤔. Internet says it should be `&nbsp;`

Comment (qurben) [reply 03-build-analytics.Rmd]: ~ just becomes ~

File: 03-build-analytics.Rmd [hunk 5]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+  * predicting build failures @santolucito2018statically, @ni2018acona
+
+The tools that are being proposed are:
+
+  * BART to help developers fix build errors by generating a summary of the failures
+  with useful information, thus eliminating the need to browse error logs @vassallo2018break
+  * HireBuild to automatically fix build failures based on previous changes @hassan2018hirebuild
+  * VeriCI capable of checking the errors in CI configurations files before the developer 
+  push a commit and without needing to wait for the build result @santolucito2018statically
+  * ACONA capable of predicting build failure prediction in CI environment for newer projects
```
Comment (Bjorn48) [inline 03-build-analytics.Rmd]: It says that ACONA predicts predictions here. Think that is not intended. :p

File: 03-build-analytics.Rmd [hunk 6]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+newer projects with less training data available.
+
+The papers identified using the research protocol defined in the previous section that give as an
+overview of the current state of the art in build analytics domain are:
+* Hassan et al. 2018 
+* Vassallo et al. 2018
+* Baltes et al. 2018
+* Bisong et al. 2017
+* Santolucito et al. 2018
+* Ni et al. 2018
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit 23f0b14245c6ea96d4b38e9bfc5ea07bc9dd9840.

File: 03-build-analytics.Rmd (hunk 5 above)
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit 376119a649321d8db2d9c2ff170358cfa0b97a90.

File: 03-build-analytics.Rmd [hunk 7]
```diff
@@ -9,30 +9,52 @@ and many other problems. This chapter is aimed to give an overview of the effort
...
 
+<!-- The following paragraph is prone to errors when auto formatting. After a ..^[..] or ..^[..],
+there _must_ be a space, else it is interpreted as superscript. --> 
+
+Build analytics covers research done on data extracted from a build process inside a project. This
+contains amongst others, build logs from continuous integration such as Travis&nbsp;CI^[See
+https://travis-ci.org/], Circle&nbsp;CI^[See https://circleci.com/], Jenkins^[See https://jenkins.io/], 
+AppVeyor^[See\ https://www.appveyor.com/] and TeamCity^[See https://www.jetbrains.com/teamcity/] or
+surveys among developers about their usage of Continuous Integration or build systems. This
+information is often paired with data from Version Control Systems (VCS) such as git.
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit fa4b5194f8b82773210a9ba1cb6294adb6443278.

File: 03-build-analytics.Rmd [hunk 8]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+useful on one hand for software developers and project managers for a better time management scheme 
+and one the other hand for other researchers that may improve their proposed models. 
+
+Moreover, the usage of automation build tools introduces a delay in the development cycle generated by
+the waiting time until the build finish successfully. One of the most recent papers analyzed 
+@santolucito2018statically presents a tool VeriCI capable of checking the errors in CI configurations 
+files before the developer push a commit and without needing to wait for the build result. This paper
+focuses on prediction of build failure without using metadata like number of commits, code churn also 
+in the learning process, but relying on the actual user programs and configuration scripts. This fact 
+makes the identification of the error cause possible. VeriCI achieve 83% accuracy of predicting build
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit 19ec5fbe732ab54517fd21e799fabd2c85296180.

File: 03-build-analytics.Rmd [hunk 9]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+that were able to predict the build taking advantage of the 56 features presented in TravisTorrent 
+build records. There models performed quite well with an R-Squared of around 80%, meaning that they
+were able to capture the variation of build time over multiple projects. Their research could be 
+useful on one hand for software developers and project managers for a better time management scheme 
+and one the other hand for other researchers that may improve their proposed models. 
+
+Moreover, the usage of automation build tools introduces a delay in the development cycle generated by
+the waiting time until the build finish successfully. One of the most recent papers analyzed 
+@santolucito2018statically presents a tool VeriCI capable of checking the errors in CI configurations 
+files before the developer push a commit and without needing to wait for the build result. This paper
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit 7883a38c282a54a502006b76b051e98d74721d21.

File: 03-build-analytics.Rmd [hunk 10]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+| 17. Ni et al.           | GScholar Search     | @ni2018acona               |  RQ1 ||
+
 
 ## Answers
 
-<!-- Group per RQ -->
+**RQ1**: What is the current state of the art in the field of build analytics?
 
-Through this we found the following papers
+The current state-of-the-art in build analytics domain refers to the use of machine learning techniques
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 11]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+to increase the productivity when using Continuos Integration (CI), to generate constraints on the 
+configuration of the CI that could improve build success rate and to predict build failures even for 
+newer projects with less training data available.
+
+The papers identified using the research protocol defined in the previous section that give as an
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 12]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+  * predicting build time @bisong2017built
+  * predicting build failures @santolucito2018statically, @ni2018acona
+
+The tools that are being proposed are:
+
+  * BART to help developers fix build errors by generating a summary of the failures
+  with useful information, thus eliminating the need to browse error logs @vassallo2018break
+  * HireBuild to automatically fix build failures based on previous changes @hassan2018hirebuild
+  * VeriCI capable of checking the errors in CI configurations files before the developer 
+  push a commit and without needing to wait for the build result @santolucito2018statically
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 13]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+  * ACONA capable of predicting build failure prediction in CI environment for newer projects 
+  with less data available @ni2018acona
+
+The build process is important part of a project that uses VCS in the way that based on the findings
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 14]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+from @hassan2018hirebuild 22% of code commits include changes in build script files for either
+build working or build fixing purposes. Moreover, recent studies have focused on how satisfied are 
+the users of CI tools, one particular paper @widder2018m analyzing what factors have an impact on
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 15]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+abandonment of Travis&nbsp;CI. The paper finds that increased build complexity reduces the chance 
+of abandonment, but larger projects abandon at a higher rate and that a project's language has 
+significant but varying effect. A surprising result is that metrics of configuration attempts 
+and knowledge dispersion in the project do not affect the rate of abandonment.
+
+
+Another important aspect is the impact of CI in the development process efficiency. On of the
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 16]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+papers that adresses this matters is @bisong2017built. This paper aims to find a balance between
+the frequency of integration and developer's productivity by proposing a machine learning models
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 17]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+that were able to predict the build taking advantage of the 56 features presented in TravisTorrent 
+build records. There models performed quite well with an R-Squared of around 80%, meaning that they
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

File: 03-build-analytics.Rmd [hunk 18]
```diff
@@ -44,15 +66,193 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+were able to capture the variation of build time over multiple projects. Their research could be 
+useful on one hand for software developers and project managers for a better time management scheme 
+and one the other hand for other researchers that may improve their proposed models.
```
[the remark this answers is not in the document]
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Fixed by Commit d98913e44dc7d05f39998aa6e687fdb9a0351ba9.

**Comment by mpsijm** [comment]
Since we (Awesome Analytics Agency) need to review your chapter, can you indicate when you are done making changes? Then we will wait with printing your chapter until then :slightly_smiling_face: 

(CC @jessetilro @nsalminen)

File: 03-build-analytics.Rmd [hunk 19]
```diff
@@ -44,214 +72,316 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
-<!-- Note: Interesting collab between Nanjing (China), Eindhoven (Netherlands), UC Davis (USA) and Carnegie Mellon (USA). -->
+The current state-of-the-art in the build analytics domain refers to the use of machine learning techniques
+to increase the productivity when using Continuos Integration (CI), to generate constraints on the 
+configuration of the CI that could improve build success rate and to predict build failures even for 
+newer projects with less training data available.
+
+The papers identified using the research protocol defined in the previous section that give us an
+overview of the current state of the art in build analytics domain are:
+
+ * Hassan et al. 2018
```
Comment (qurben) [inline 03-build-analytics.Rmd]: Please use reference notation

File: 03-build-analytics.Rmd [hunk 20]
```diff
@@ -44,214 +72,316 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
 
-This paper explores which CI system developers use, how developers use CI and why developers use
-CI. For this it analyzes data from Github, Travis CI and it conducts a developer survey. It finds
-that projects using CI release twice as often, accept pull requests faster and have developers who
-are less worried about breaking the build.
+The papers identified using the research protocol defined in the previous section that give us an
+overview of the current state of the art in build analytics domain are:
 
-### @vassallo2017tale
+ * Hitlon et al. 2016
```
Comment (qurben) [inline 03-build-analytics.Rmd]: Please use citation notation

File: 03-build-analytics.Rmd [hunk 21]
```diff
@@ -44,214 +72,316 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
+![CI overview.](figures/Chapter3/state_pr.png)
 
-_Initial Seed_
 
-This paper explores which CI system developers use, how developers use CI and why developers use
-CI. For this it analyzes data from Github, Travis CI and it conducts a developer survey. It finds
-that projects using CI release twice as often, accept pull requests faster and have developers who
-are less worried about breaking the build.
+The papers identified using the research protocol defined in the previous section that give us an
+overview of the current state of the art in build analytics domain are:
```
Comment (qurben) [inline 03-build-analytics.Rmd]: _field_ or _domain_?

_the_ build analytics

File: 03-build-analytics.Rmd [hunk 22]
```diff
@@ -44,214 +72,316 @@ The keywords used were: build analytics, machine learning, build time, predictio
...
-_Initial Seed_
-
-<!-- Note: Interesting collab between Nanjing (China), Eindhoven (Netherlands), UC Davis (USA) and Carnegie Mellon (USA). -->
+The current state-of-the-art in the build analytics domain refers to the use of machine learning techniques
+to increase the productivity when using Continuos Integration (CI), to generate constraints on the 
+configuration of the CI that could improve build success rate and to predict build failures even for 
+newer projects with less training data available.
+
+The papers identified using the research protocol defined in the previous section that give us an
+overview of the current state of the art in build analytics domain are:
```
Comment (qurben) [inline 03-build-analytics.Rmd]: https://github.com/saltudelft/software-analytics-book/pull/19/files#r223225271

File: 03-build-analytics.Rmd (hunk 19 above)
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Resolved by Commit 1b9af366c953aa0e53c4215ac4870c2f982d7a31.

File: 03-build-analytics.Rmd (hunk 20 above)
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Resolved by Commit 77510941a910c400fc4e1168713e0340480e3662.

File: 03-build-analytics.Rmd (hunk 21 above)
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Resolved by Commit 0d500095d68629c4d69d9bfb4d192a7193f8d8f5.

File: 03-build-analytics.Rmd (hunk 22 above)
Comment (simionAndrei) [reply 03-build-analytics.Rmd]: Resolved by Commit 0d50009.

**Comment by qurben** [comment]
Latest version of our book is found on the following site:
https://chart-increasing-keyboard-analytics.github.io/software-analytics-book/build-analytics.html

PDF is also found there and at:
https://chart-increasing-keyboard-analytics.github.io/software-analytics-book/software-analytics-survey.pdf

**Comment by qurben** [comment]
@mpsijm You can press the print button!

File: 03-build-analytics.Rmd [hunk 23]
```diff
@@ -2,256 +2,422 @@
...
+
+### Build Definition
+Build analytics covers research done on data extracted from a build process inside a project. This
+contains among others, build logs from Continuous Integration such as Travis&nbsp;CI^[See
+https://travis-ci.org/], Circle&nbsp;CI^[See https://circleci.com/], Jenkins^[See https://jenkins.io/], 
+AppVeyor^[See\ https://www.appveyor.com/] and TeamCity^[See https://www.jetbrains.com/teamcity/] or
+surveys among developers about their usage of Continuous Integration or build systems. This
+information is often paired with data from Version Control Systems such as Git. 
 
 <!--
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: Is this comment still needed? (Similar for other comments later, though I must say I like that you used them to structure your work!)

File: 03-build-analytics.Rmd [hunk 24]
```diff
@@ -2,256 +2,422 @@
...
+Continuous Integration in a term used in software engineering to describe a practice of merging 
+all developer working copies to a shared mainline several times a day. CI is in general used 
+together with Version Control System (VCS), an application for revision control that ensures 
+the management of changes to documents, source code and other collections of information.
+
+<!-- The following paragraph is prone to errors when auto formatting. After a ..^[..] or ..^[..],
+there _must_ be a space, else it is interpreted as superscript. --> 
+
+### Build Definition
+Build analytics covers research done on data extracted from a build process inside a project. This
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: isn't "done" redundant?

File: 03-build-analytics.Rmd [hunk 25]
```diff
@@ -2,256 +2,422 @@
...
+failed test or inspection, problem with the database or failed deployment. In the CI environment, it is 
+important that these problems are fixed immediately. 
 
-### @fowler2006continuous
+#### Keep the Build Fast
 
-In this paper, Martin talks about the current state of the software industry in terms of Continuous Integration (CI) and comments on the practises required to implement CI effectively. He talks about his experience working for a large English electronics company where the development of a project took two years and the integration process took several months. Integration is a long and unpredictable process. Martin suggested this approach and that the two most common reactions he got were: “it can’t work (here)” or “doing it won’t make much difference”.  He expresses that most engineers don’t know how simple the process can be of setting the CI framework up. In this way, we get a glimpse into the practises popular within the industry regarding build analytics.
+The build should complete rapidly, so if there is an issue with integration, it can be identified quickly.
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: Where does this statement come from?

File: 03-build-analytics.Rmd [hunk 26]
```diff
@@ -2,256 +2,422 @@
...
+
+Tools presented in papers might require a more large scale and long term study to verify that the
+tool presented keeps up when it is actually used[@santolucito2018statically].
+
+Future research in build analytics branches in a couple of different topics. Pinto and Rebouças[@pinto2018work]
+proposes to focus on getting a better understanding of the users and why they might choose to
+abandon an automatic build platform.
 
-### @stolberg2009enabling
+Baltes et al.[@baltes2018no] suggests that in future research more perspectives when analyzing commit data should
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: How about
"Baltes et al. [@baltes2018no] suggest" ?

Please consolidate with rest of the document

File: 03-build-analytics.Rmd [hunk 27]
```diff
@@ -2,256 +2,422 @@
...
+surveys among developers about their usage of Continuous Integration or build systems. This
+information is often paired with data from Version Control Systems such as Git. 
 
 <!--
  - Define build
  - Define context
  -->
 
-## Research Questions
+### Research Questions
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: How about at least mentioning the RQs directly in the motivation, as the other groups have done (see https://saltudelft.github.io/software-analytics-book/testing-analytics.html for example)?

File: 03-build-analytics.Rmd [hunk 28]
```diff
@@ -2,256 +2,422 @@
...
+questions based on the open research and challenges in current research.
+
+## Research Protocol {#build-analytics-research-protocol}
+
+### Search Strategy
+Taking advantage of the initial seed consisting of Bird and Zimmermann [@bird2017predicting],
+Beller et al. [@beller2017oops], Rausch et al. [@rausch2017empirical], Beller et al. TravisTorrent [@beller2017travistorrent],
+Pinto et al. [@pinto2018work], Zhao et al. [@zhao2017impact], Widder et al. [@widder2018m] and
+Hilton et al. [@hilton2016usage], we used references to find new papers to analyze.
+Moreover, we used academical search engines like _Google_ _Scholar_ to perform a keyword based
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: Please be more specific: which search engine did use exactly?

Comment (Inventitech) [inline 03-build-analytics.Rmd]: keyword-based

File: 03-build-analytics.Rmd (hunk 27 above)
Comment (qurben) [reply 03-build-analytics.Rmd]: This is inside te motivation? Or is this a comment on the maybe a bit overzealous sectioning?

**Pull request merged**